### PR TITLE
feat(datepicker): add quick select mode to date range picker (#5262)

### LINKED ISSUE #5262: "Quick select" date ranges for date range picker
### Bug description or feature request:

Feature Request.

I'd really like to be able to achieve a similar effect to http://www.daterangepicker.com/ and https://www.npmjs.com/package/ngx-daterangepicker-material where you can set custom "quick select" ranges for the date when the picker is in range mode. If that functionality could be added to ngx-bootstrap's date range picker that would be awesome, however, my fallback plan would have been to use an inline date range picker nested within a Popover. That would have worked, except there's no inline date range picker, only an inline date picker, if the "quick select" ranges don't happen, second on my wish list would be the ability to inline a date range picker.

### Versions of ngx-bootstrap, Angular, and Bootstrap:

ngx-bootstrap: 4.2.0

Angular: 8.0.0

Bootstrap: 4.3.1

### LINKED ISSUE #5542: Adding Date Presets to DateRangePicker
**Feature Request:**
Adding multiple preset date ranges as a capability to the library 

**Probable solutions**

Adding support as a configuration in bsConfig
**OR**
Creating a new "preset" property

**Workarounds if any**
Current workaround requires me to write the following code which basically renders the panel on the right of the date container and the click functions update the model for the input element

image

**html**
```
<div class="input-group">
              <input
                type="text"
                #dp="bsDaterangepicker"
                placeholder="Select Date Range"
                class="form-control bs-datepicker-input"
                bsDaterangepicker
                [minDate]="model.minDate"
                [maxDate]="model.maxDate"
                [bsConfig]="model.bsConfig"
                [(ngModel)]="model.value"
                (ngModelChange)="
                  getDataWithinRange(
                    model,
                    dataList
                  )
                "
              />
              <div
                class="offset-md-3 col-md-6 custom-date-select"
                *ngIf="dp.isOpen"
              >
                <div
                  class="row col"
                  *ngFor="let datePreset of model.datePresetOptions"
                >
                  <div
                    class="col-md-12 text-center"
                    (click)="
                      getDataWithinRange(
                        model,
                        dataList,
                        datePreset.key
                      )
                    "
                  >
                    <a
                      [ngClass]="
                        selectedDatePreset === datePreset.key ? 'selected' : ''
                      "
                    >
                      {{ datePreset.label }}
                    </a>
                  </div>
                </div>
              </div>
            </div>
```

**css**
```
.custom-date-select {
  position: absolute;
  z-index: 1081;
  left: 24rem;
  top: 2rem;
  cursor: pointer;
  align-self: center;
  width: 100%;
  padding: 1em 3em 0 0;


  .row {
    div {
      padding: .5em;
      border-bottom: 1px solid $THEME-COLOR-WHITE-7;

      a {
        color: $THEME-COLOR-GREEN-2;

        &.selected {
          color: $THEME-COLOR-BLUE-1;
        }
      }
    }
  }
}
```

**JS**
```
this.model = {
      minDate: new Date(0),
      maxDate: new Date(),
      bsConfig: {
        isAnimated: true,
        dateInputFormat: 'DD--MM-YYYY',
        containerClass: 'data-grid-bs-datepicker'
      },
      value: []
}
```

I would rather want all this to be a part of the component and developers have this capability as easy as writing

```<input [preset]=DATE_PRESET [bsConfig]="{presetLocation: left / right / bottom }"/>```

```
const DATE_PRESET_KEYS = {
  TODAY: 'TODAY',
  LAST_7_DAYS: 'LAST_7_DAYS',
  LAST_30_DAYS: 'LAST_30_DAYS',
  LAST_60_DAYS: 'LAST_60_DAYS',
  LAST_90_DAYS: 'LAST_90_DAYS'
};

const DATE_PRESET = [
  {
    label: 'TODAY',
    key: DATE_PRESET_KEYS.TODAY
  },
  {
    label: 'Last 7 DAYS',
    key: DATE_PRESET_KEYS.LAST_7_DAYS
  },
  {
    label: 'Last 30 Days',
    key: DATE_PRESET_KEYS.LAST_30_DAYS
  },
  {
    label: 'Last 60 Days',
    key: DATE_PRESET_KEYS.LAST_60_DAYS
  },
  {
    label: 'Last 90 Days',
    key: DATE_PRESET_KEYS.LAST_90_DAYS
  }
];
```

### PR DESCRIPTION
**Pull request by arvindm95**
Closes #5262
closes #5542
# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated tests.
 - [x] added/updated API documentation.
 - [x] added/updated demos.

### CODE REVIEW

**Review by ghost — DISMISSED** [review]

**Review by daniloff200 — DISMISSED** [review]

**Review by daniloff200 — DISMISSED** [review]

File: src/datepicker/bs-daterangepicker-inline.component.ts [hunk 1]
```diff
@@ -55,6 +56,10 @@ export class BsDaterangepickerInlineDirective implements OnInit, OnDestroy, OnCh
      * Disable specific dates
      */
     @Input() datesDisabled: Date[];
+    /**
+     * Predefined ranges
+     */
+    @Input() ranges: BsCustomDates[];
```
Comment (valorkin) [inline src/datepicker/bs-daterangepicker-inline.component.ts]: please push this to component via `config`, let's not bloat components API
thanks in advance!

File: src/datepicker/bs-daterangepicker-inline.component.ts [hunk 2]
```diff
@@ -145,7 +150,8 @@ export class BsDaterangepickerInlineDirective implements OnInit, OnDestroy, OnCh
         minDate: this.minDate || this.bsConfig && this.bsConfig.minDate,
         maxDate: this.maxDate || this.bsConfig && this.bsConfig.maxDate,
         dateCustomClasses: this.dateCustomClasses || this.bsConfig && this.bsConfig.dateCustomClasses,
-        datesDisabled: this.datesDisabled || this.bsConfig && this.bsConfig.datesDisabled
+        datesDisabled: this.datesDisabled || this.bsConfig && this.bsConfig.datesDisabled,
+        ranges: this.ranges
```
Comment (valorkin) [inline src/datepicker/bs-daterangepicker-inline.component.ts]: add support for `this.bsConfig`

File: src/datepicker/bs-daterangepicker.component.ts [hunk 3]
```diff
@@ -110,6 +111,12 @@ export class BsDaterangepickerDirective
    * Disable specific dates
    */
   @Input() datesDisabled: Date[];
+
+  /**
+   * Predefined ranges
+   */
+  @Input() ranges: BsCustomDates[];
```
Comment (valorkin) [inline src/datepicker/bs-daterangepicker.component.ts]: via `config` please

File: src/datepicker/bs-daterangepicker.component.ts [hunk 4]
```diff
@@ -223,7 +230,8 @@ export class BsDaterangepickerDirective
         minDate: this.minDate || this.bsConfig && this.bsConfig.minDate,
         maxDate: this.maxDate || this.bsConfig && this.bsConfig.maxDate,
         dateCustomClasses: this.dateCustomClasses || this.bsConfig && this.bsConfig.dateCustomClasses,
-        datesDisabled: this.datesDisabled || this.bsConfig && this.bsConfig.datesDisabled
+        datesDisabled: this.datesDisabled || this.bsConfig && this.bsConfig.datesDisabled,
+        ranges: this.ranges
```
Comment (valorkin) [inline src/datepicker/bs-daterangepicker.component.ts]: add support for `this.config`

File: src/datepicker/bs-daterangepicker.spec.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
+import { Component, ViewChild } from '@angular/core';
+import {
+    BsDaterangepickerDirective,
+    BsDaterangepickerConfig,
+    BsDatepickerModule,
+    BsDaterangepickerContainerComponent
+} from '.';
```
Comment (valorkin) [inline src/datepicker/bs-daterangepicker.spec.ts]: please don't import from `.`, please import from specific files

File: src/datepicker/themes/bs/bs-daterangepicker-container.component.ts [hunk 6]
```diff
@@ -49,6 +50,8 @@ export class BsDaterangepickerContainerComponent extends BsDatepickerAbstractCom
     super();
     this._effects = _effects;
 
+    this._customRangesFish = this._config.ranges;
```
Comment (valorkin) [inline src/datepicker/themes/bs/bs-daterangepicker-container.component.ts]: why fish? plus this `_customRangesFish` field is public, so it should be without underscore

**Review by valorkin — DISMISSED** [review]
in general looks good, please apply small changes requested

File: src/datepicker/bs-daterangepicker-inline.component.ts (hunk 1 above)
Comment (arvindm95) [reply src/datepicker/bs-daterangepicker-inline.component.ts]: Yes, removed this and moved this to load from config.

File: src/datepicker/bs-daterangepicker-inline.component.ts (hunk 2 above)
Comment (arvindm95) [reply src/datepicker/bs-daterangepicker-inline.component.ts]: Done

File: src/datepicker/bs-daterangepicker.component.ts (hunk 3 above)
Comment (arvindm95) [reply src/datepicker/bs-daterangepicker.component.ts]: Done

File: src/datepicker/bs-daterangepicker.component.ts (hunk 4 above)
Comment (arvindm95) [reply src/datepicker/bs-daterangepicker.component.ts]: Done

File: src/datepicker/bs-daterangepicker.spec.ts (hunk 5 above)
Comment (arvindm95) [reply src/datepicker/bs-daterangepicker.spec.ts]: Hi, I have changed it in this file, but there are few other spec files which import from `.`. Maybe going forward we need to modify those files as well. Thanks!

File: src/datepicker/themes/bs/bs-daterangepicker-container.component.ts (hunk 6 above)
Comment (arvindm95) [reply src/datepicker/themes/bs/bs-daterangepicker-container.component.ts]: Done.

File: src/datepicker/bs-daterangepicker.spec.ts (hunk 5 above)
Comment (daniloff200) [reply src/datepicker/bs-daterangepicker.spec.ts]: Yeah, will do that

**Comment by arvindm95** [comment]
@valorkin Done with the changes, can you please take a look. Thanks!

**Review by IraErshova — APPROVED** [review]

**Comment by IraErshova** [comment]
@dmitry-zhemchugov Could you please test this PR?

**Comment by daniloff200** [comment]
@IraErshova that was tested by Dima longtime ago but was blocked by Valorkin  :)
also there was some comments from him, but looks like that was fixed. 

I'm gonna review all things today again and test it by myself too, I think we can merge2 it, besides , our folks from R*** project need that feature too

**Review by daniloff200 — APPROVED** [review]

**Pull request merged**